File: backend/app/utils/district_filter.py
"""İlçe adı eşleştirme — Google/DB sonuçlarını semte göre filtrele."""

from __future__ import annotations

import unicodedata


def _norm(text: str) -> str:
    t = unicodedata.normalize('NFKD', text.strip().lower())
    return ''.join(c for c in t if not unicodedata.combining(c))


def district_name_variants(district: str) -> set[str]:
    raw = district.strip()
    if not raw:
        return set()
    n = _norm(raw)
    return {raw, raw.lower(), n, n.replace(' ', '')}


def address_matches_district(address: str, district: str) -> bool:
    if not district.strip():
        return True
    addr = _norm(address or '')
    if not addr:
        return False
    for variant in district_name_variants(district):
        v = _norm(variant)
        if len(v) >= 3 and v in addr:
            return True
    return False


def filter_by_district(items: list, district: str, *, address_attr: str = 'address') -> list:
    if not district.strip():
        return items
    out = []
    for item in items:
        addr = getattr(item, address_attr, '') or ''
        name = getattr(item, 'name', '') or ''
        if address_matches_district(addr, district) or address_matches_district(name, district):
            out.append(item)
    return out

File: frontend/src/utils/district-filter.ts
/** İlçe adı eşleştirme — Google sonuçlarını semte göre filtrele */

function norm(text: string): string {
  return text
    .toLowerCase()
    .replace(/ı/g, 'i')
    .replace(/ğ/g, 'g')
    .replace(/ü/g, 'u')
    .replace(/ş/g, 's')
    .replace(/ö/g, 'o')
    .replace(/ç/g, 'c')
    .trim();
}

export function addressMatchesDistrict(address: string, district: string): boolean {
  if (!district.trim()) return true;
  const addr = norm(address);
  const d = norm(district);
  if (addr.includes(d) || d.includes(addr)) return true;
  const parts = d.split(/\s+/).filter((p) => p.length >= 3);
  if (parts.length > 0 && parts.every((part) => addr.includes(part))) return true;
  const first = parts[0];
  if (first && first.length >= 4 && addr.includes(first)) return true;
  return false;
}

export function filterGoogleByDistrict<T extends { address?: string; name?: string }>(
  items: T[],
  district: string,
): T[] {
  if (!district.trim()) return items;
  return items.filter(
    (p) => addressMatchesDistrict(p.address ?? '', district) || addressMatchesDistrict(p.name ?? '', district),
  );
}

export function filterGoogleByCity<T extends { address?: string; name?: string }>(
  items: T[],
  city: string,
): T[] {
  if (!city.trim()) return items;
  const c = norm(city);
  return items.filter((p) => {
    const addr = norm(p.address ?? '');
    const name = norm(p.name ?? '');
    return addr.includes(c) || name.includes(c) || c.split(/\s+/).every((part) => part.length >= 3 && addr.includes(part));
  });
}
